gpiozero/pins/pigpiod.py

```diff
@@ -259,20 +277,224 @@ def _set_when_changed(self, value):
             self._callback.cancel()
             self._callback = None
         if value is not None:
-            self._callback = self._connection.callback(
-                    self._number, self._edges,
+            self._callback = self.factory.connection.callback(
+                    self.number, self._edges,
                     lambda gpio, level, tick: value())
 
-    @classmethod
-    def pi_info(
-            cls, host=os.getenv('PIGPIO_ADDR', 'localhost'),
-            port=int(os.getenv('PIGPIO_PORT', 8888))):
+
+class PiGPIOHardwareSPI(SPI, Device):
+    def __init__(self, factory, port, device):
+        self._port = port
+        self._device = device
+        self._factory = weakref.proxy(factory)
+        super(PiGPIOHardwareSPI, self).__init__()
+        self._reserve_pins(*(
+            factory.address + ('GPIO%d' % pin,)
+            for pin in (11, 10, 9, (8, 7)[device])
+            ))
+        self._mode = 0
+        self._select_high = False
+        self._bits_per_word = 8
+        self._baud = 500000
+        self._handle = self._factory.connection.spi_open(
+            device, self._baud, self._spi_flags())
+
+    def close(self):
         try:
-            connection, info = cls._CONNECTIONS[(host, port)]
-        except KeyError:
-            connection = pigpio.pi(host, port)
-            revision = '%04x' % connection.get_hardware_revision()
-            info = pi_info(revision)
-            cls._CONNECTIONS[(host, port)] = (connection, info)
-        return info
+            self._factory._spis.remove(self)
+        except (ReferenceError, ValueError):
+            # If the factory has died already or we're not present in its
+            # internal list, ignore the error
+            pass
+        if not self.closed:
+            self._factory.connection.spi_close(self._handle)
+        self._handle = None
+        self._release_all()
+        super(PiGPIOHardwareSPI, self).close()
+
+    @property
+    def closed(self):
+        return self._handle is None or self._factory.connection is None
+
+    @property
+    def factory(self):
+        return self._factory
+
+    def __repr__(self):
+        try:
+            self._check_open()
+            return 'SPI(port=%d, device=%d)' % (self._port, self._device)
+        except DeviceClosed:
+            return 'SPI(closed)'
+
+    def _spi_flags(self):
+        return (
+            self._mode          << 0                  |
+            self._select_high   << (2 + self._device) |
+            self._bits_per_word << 16
+            )
+
+    def _get_clock_mode(self):
+        return self._clock_mode
+
+    def _set_clock_mode(self, value):
+        self._check_open()
+        if not 0 <= value < 4:
+            raise SPIInvalidClockmode("%d is not a valid SPI clock mode" % value)
```

Comment:
Seems like this identical check occurs in multiple places - I wonder if it could be moved to some parent class?